src/lib/formats/dmk_dsk.cpp

```diff
@@ -95,117 +114,207 @@ bool dmk_format::load(util::random_read &io, uint32_t form_factor, const std::ve
+			bool idam_is_mfm = (track_data[track_header_offset + 1] & 0x80) ? true : false;
 			track_header_offset += 2;
 
 			while ( track_offset != 0 && track_offset >= 0x83 && track_offset < track_size && track_header_offset < 0x80 )
 			{
 				// Assume 3 bytes before IDAM pointers are the start of IDAM indicators
-				idam_location[(track_header_offset/2) - 1] = track_offset - 3;
+				int mark_offset = idam_is_mfm ? 3 : 0;
+				mark_location[mark_count] = track_offset - mark_offset;
+				mark_value[mark_count] = 0xfe;
+				mark_is_mfm[mark_count] = idam_is_mfm;
+				mark_count++;
 
+				int stride = idam_is_mfm ? 1 : fm_stride;
 				// Scan for DAM location
-				for (int i = track_offset + 53; i > track_offset + 10; i--)
+				for (int i = track_offset + 10*stride; i < track_offset + 53*stride; i++)
 				{
-					if ((track_data[i] == 0xfb || track_data[i] == 0xf8) && track_data[i-1] == 0xa1 && track_data[i-2] == 0xa1)
+					if ((track_data[i] >= 0xf8 && track_data[i] <= 0xfb))
 					{
-						dam_location[(track_header_offset/2) - 1] = i - 3;
-						break;
+						if (!idam_is_mfm || (track_data[i-1] == 0xa1 && track_data[i-2] == 0xa1))
+						{
+							mark_location[mark_count] = i - mark_offset;
+							mark_value[mark_count] = track_data[i];
+							mark_is_mfm[mark_count] = idam_is_mfm;
+							mark_count++;
+
+							break;
+						}
 					}
 				}
 
+				idam_is_mfm = (track_data[track_header_offset + 1] & 0x80) ? true : false;
 				track_offset = ( ( track_data[track_header_offset + 1] << 8 ) | track_data[track_header_offset] ) & 0x3fff;
 				track_header_offset += 2;
-			};
+			}
+
+			// Prevent encoding from switching after last sector
+			if (mark_count > 0)
+			{
+				mark_is_mfm[mark_count] = mark_is_mfm[mark_count - 1];
+			}
 
 			// Find IAM location
-			for(int i = idam_location[0] - 1; i >= 3; i--)
+			for(int i = mark_location[0] - 1; i >= 3; i--)
 			{
 				// It's usually 3 bytes but several dumped tracks seem to contain only 2 bytes
-				if (track_data[i] == 0xfc && track_data[i-1] == 0xc2 && track_data[i-2] == 0xc2)
+				if (track_data[i] == 0xfc && (is_sd || (track_data[i-1] == 0xc2 && track_data[i-2] == 0xc2)))
 				{
-					iam_location = i - 3;
+					iam_location = i - (is_sd ? 0 : 3);
 					break;
 				}
 			}
 
-			int idam_index = 0;
-			int dam_index = 0;
+			int curr_mark = 0;
+			bool enc_mfm = mark_is_mfm[curr_mark];
 			for (int offset = 0x80; offset < track_size; offset++)
 			{
 				if (offset == iam_location)
 				{
-					// Write IAM
-					raw_w(raw_track_data, 16, 0x5224);
-					raw_w(raw_track_data, 16, 0x5224);
-					raw_w(raw_track_data, 16, 0x5224);
-					offset += 3;
+					if (!is_sd)
+					{
+						// Write IAM
+						raw_w(raw_track_data, 16, 0x5224);
+						raw_w(raw_track_data, 16, 0x5224);
+						raw_w(raw_track_data, 16, 0x5224);
+						offset += 3;
+					}
+					else
+					{
+						raw_w(raw_track_data, 32, wide_fm(0xff7a));	// FC clocked with D7
```

Comment:
Shouldn't FC clocked with D7 be `0xf77a`?

Reply:
Good catch, yes it should be 0xf77a.